packages/synapse-react-client/src/components/SchemaDrivenAnnotationEditor/template/ObjectFieldTemplate.tsx

```diff
@@ -0,0 +1,162 @@
+  F extends FormContextType = any,
+>(props: ObjectFieldTemplateProps<T, S, F>) {
+  const {
+    description,
+    disabled,
+    formData,
+    idSchema,
+    onAddClick,
+    properties,
+    readonly,
+    registry,
+    required,
+    schema,
+    title,
+    uiSchema,
+  } = props
+
+  const options = getUiOptions<T, S, F>(uiSchema)
+  const TitleFieldTemplate = getTemplate<'TitleFieldTemplate', T, S, F>(
+    'TitleFieldTemplate',
+    registry,
+    options,
+  )
+  const DescriptionFieldTemplate = getTemplate<
+    'DescriptionFieldTemplate',
+    T,
+    S,
+    F
+  >('DescriptionFieldTemplate', registry, options)
+
+  /*
+     Tracks the property keys that were last known to exist in the object.
+     */
+  const previousSchemaDefinedProperties = useRef<Set<string>>(new Set())
+
+  /**
+   * We track how the schema changes as the user enters data, causing conditional subschemas to be evaluated.
+   *
+   * If a property exists in the previous version of the schema and has user data, and then is dropped from the schema due to a data update, then
+   * we need to prompt the user whether they want to undo the change, or continue and keep/remove those fields
+   *
+   * In this component, we can identify when one or more fields with user data are dropped (the field gains the additional property flag in the schema prop).
+   * We then use a function provided by the context to report which fields were lost.
+   */
+  useDeepCompareEffectNoCheck(() => {
+    if ('properties' in schema && schema.properties) {
+      const propertyKeys = Object.keys(schema.properties)
+      // Schema-defined properties are those properties in the schema without the additional property flag.
+      const schemaDefinedProperties = new Set<string>(
+        propertyKeys.filter(key => {
+          const propertyObject = schema.properties![key]
+          return !propertyObject[ADDITIONAL_PROPERTY_FLAG]
+        }),
+      )
+
+      if (previousSchemaDefinedProperties.current != null) {
+        // Compare the schema defined properties with the previous version to see if any were lost.
+        const lostProperties = Array.from(
+          previousSchemaDefinedProperties.current,
+        ).filter(schemaDefinedProperty => {
+          return (
+            // Property is no longer in schema
+            !schemaDefinedProperties.has(schemaDefinedProperty) &&
+            // Form data exists
+            !!formData &&
+            // Property still exists in form data and is not null
+            // if the data is null, then we don't need to worry about it; user data wasn't lost
+            formData[schemaDefinedProperty] != null &&
+            // Property matches `additionalProperties` requirements: is an array and contains non-null values
+            Array.isArray(formData[schemaDefinedProperty]) &&
+            (formData[schemaDefinedProperty] as Array<any>).filter(
+              item => item != null,
+            ).length > 0
```

Comment:
Just added these conditions to prevent notifications on cases where the data is empty and will be dropped.